### ISSUE TITLE
1.0.0 Planning

### ISSUE THREAD

**Issue by DamianEdwards** [aspnet/SignalR-Server#196]
A refresher on the high-level features SignalR offers (that may or may not be reassessed):
- Auto-reconnect of connections
- Message storage and replay on reconnect (required for Long Polling to work at all really)
- Persistent connection and Hub programming abstractions
- Transport negotiation (via fallback)
- Round-tripped server/client shared state (Hub state)
- Plug-in scale-out providers

The givens:
- Make SignalR server work on ASP.NET Core (including .NET Core)
- Must support existing clients
  - We may remove some features though
- Rewrite JS client using TypeScript and no dependencies

Some ideas:
- Try to remove some of the complexity that has eventuated in the code-base today
- Re-baseline our supported browsers
- Support pluggable serialization
- Support JSON-RPC clients
- Support binary
- Move issues from signalr/signalr tagged as v3 to here and triage
- Make SignalR server not require a client for basic pub/sub & dispatching features
- Try to connect to server with single request (for streaming transports)
- Should the new JS client support existing SignalR server?
- Reconsider the design based on the assumption that WebSockets is the normal case, and other transports may not get all features
- Rethink scale-out
- Support server-publishers, i.e. an app that connects to a SignalR server as a client but doesn’t receive any messages, just publishes them
- Rewrite the C# client using… C#

Next steps:
- Continue discussion on whether we pull out the concept of an `IMessageBus` and build SignalR programming model on top of that

**Comment by Tragetaschen** [comment]
I just love the givens. Especially losing the jQuery dependency. Count me in for an early adapter.

To ponder over your "Web Sockets is the normal case" thought:
* Drop support for Forever Frame. I've been using an EventSource polyfill (Yaffle/EventSource) for the last couple of months and IE11 / Edge work well with it. Also, this transport is the only one that isn't "zone aware" out of the box (to use an angular2 term: angular/angular#6431)
* Consider not supporting Long Polling. As far as I understand it, it's a kludge so a buffering proxy that sits in between can be supported. Wouldn't it be better if this buffer bloat could be resolved on the platforms supported by .NET Core? (see aspnet/IISIntegration#182)
* Once you are there, SSE transport and web sockets are basically equivalent if the minimum browser version is increased to IE11 + evergreen (a reasonable choice). So then it could be feasible to have Web Socket transport as the only option.
* I know I only considered browser clients. I simply don't know what the other implementations require.

**Comment by IanYates** [comment]
@Tragetaschen - good points.
One thing though, IE 11 runs on Windows 7 but without websocket support, so you couldn't say you're doing IE11+ without the extra caveat of Win 8.1+ too.
As someone writing apps used in some enterprises, it's nice (well, necessary actually) for SignalR can still work through some firewalls that may not play nicely with websockets.

:+1: for some better way of doing binary
I'd also be keen to be able to avoid the need to round-trip the subscribed groups if possible.  Even if it reduced resiliency (I only have one server as we're an on-prem app) that'd be splendid.  I appreciate I may be in the minority with only having a single server per deployment.

**Comment by muratg** [comment]
- Investigate if we can create TypeScript declerations files for hubs on build or post build steps

**Comment by Worthaboutapig** [comment]
Can you please clarify how this project for .NET Core relates to the original project https://github.com/SignalR/SignalR?  Numerous questions have been asked there as to what's going on and no-one's giving an answer.  It seems that everyone is looking at .NET Core as the new shiny thing and just dropping everything else.  Not everyone can or wants to move to .NET Core currently.  There are 19 pull requests on the other project, many of which seem to only need merging- they've been approved.  There's hundreds of issues and some activity this month, but nothing for 6 months previously.

.NET Core is very nice, but what are people who rely on mature, production code supposed to do? Is Microsoft actually committed to both versions, or just this one?  If they're only interested in .NET Core, can you pass it over to the community so that progress can be made?

Sorry, but this has all got a bit frustrating.

**Comment by BrannonKing** [comment]
Having used SignalR in a C# client/server application for the last few years, here are my thoughts on the overall architecture for it:

1. Ditch the global state (`GlobalHost`). Everyone uses IoC frameworks now. The SignalR globals make unit testing difficult. The built-in dependency resolver has a funky lifetime because of it.

2. The non-websocket transports are unusable for me from a performance standpoint. They scan every character. My software literally runs 10x slower on Windows 7. It can't keep up. I'd say ditch those. 

3. However, I wouldn't mind the ability to swap out the transport layer for something custom. It would be nice if I could insert ZeroMQ/NetMQ or some DDS framework in place of WebSockets.

4. Break apart large messages. The current framework allows you to send large messages that clog up the pipe. That should be impossible. It might also be nice to flag a message as high-priority.

5. The .NET client required me to write my own reflection-based wrapper to make my client-side use the same interface as the server. WampSharp does that automatically, although I still don't like how WampSharp makes you separate methods and events.

6. I like the JSON-RPC plan. I had to dig around online and in the code to pull out the actual message format. 

7. Cosider using run-time compilation for the message wrappers. It was what I was playing with here: https://github.com/BrannonKing/Kts.Remoting . I've also attached my results from that study.
[SocketBenchmarks.xlsx](https://github.com/aspnet/SignalR-Server/files/376831/SocketBenchmarks.xlsx)

**Comment by RichiCoder1** [comment]
@BrannonKing A few comments on a few of those:
>2. The non-websocket transports are unusable for me from a performance standpoint. They scan every character. My software literally runs 10x slower on Windows 7. It can't keep up. I'd say ditch those.

Maybe have the option to shut off fallbacks or plug in so can record/alert when a client fallsback, but I don't think they should be removed wholesale. 

>3. However, I wouldn't mind the ability to swap out the transport layer for something custom. It would be nice if I could insert ZeroMQ/NetMQ or some DDS framework in place of WebSockets.

While an interesting idea, that seems very much to depart from the core mission of SignalR, namely to be a broadly-compatible WebSockets/Web-compatible real-time server.

>8. Any exception thrown by the framework needs to be catchable in my code. I still have a variety of SignalR / Websocket exceptions that show up in my unhandled exception catcher.

I imagine this will happen for "free" if it's built on AspNet bits since AspNet comes will logging out of the box now.

**Comment by JanEggers** [comment]
@DamianEdwards:  whats that IMessageBus interface u are referring to?

i would prefere an api that relies more on IObservable / IObserver that would bring great possibilities when combined with rx.

that way a lot of the discovery stuff could be dropped as someone could just subscribe to events on a hubcontext or publish events via hubcontext and the Interface is the type of the event

**Comment by davidfowl** [comment]
@Tragetaschen @Worthaboutapig 

> Drop support for Forever Frame. I've been using an EventSource polyfill (Yaffle/EventSource) for the last couple of months and IE11 / Edge work well with it. Also, this transport is the only one that isn't "zone aware" out of the box (to use an angular2 term: angular/angular#6431)

We're going to be doing this because of ties to the DOM and the issues it has in various versions of IE.

> Consider not supporting Long Polling. As far as I understand it, it's a kludge so a buffering proxy that sits in between can be supported. Wouldn't it be better if this buffer bloat could be resolved on the platforms supported by .NET Core? (see aspnet/IISIntegration#182)

Can't do it, it's one of the most popular transports and it fundamentally affects the overall design (stateful vs stateless) of SignalR.

> for some better way of doing binary

We care about allowing more than JSON. It'll be one of the big changes we invest in.

> I'd also be keen to be able to avoid the need to round-trip the subscribed groups if possible. Even if it reduced resiliency (I only have one server as we're an on-prem app) that'd be splendid. I appreciate I may be in the minority with only having a single server per deployment.

We can do a allow moving all state to be in memory instead of being round tripped for these types of scenarios.

**Comment by davidfowl** [comment]
@Worthaboutapig We'll still be investing in SignalR/SignalR. It will get bug fixes but no new features. New feature work and will continue in SignalR-Server for ASP.NET Core (which runs on both .NET Framework and .NET Core).

**Comment by davidfowl** [comment]
@BrannonKing 

> Ditch the global state (GlobalHost). Everyone uses IoC frameworks now. The SignalR globals make unit testing difficult. The built-in dependency resolver has a funky lifetime because of it.

It's already gone in SignalR-Server. We have a unified DI system that we are taking advantage of in SignalR Core.

> The non-websocket transports are unusable for me from a performance standpoint. They scan every character. My software literally runs 10x slower on Windows 7. It can't keep up. I'd say ditch those.


Not sure what you mean by scanning every character but we're going to be keeping long polling and SSE.

> However, I wouldn't mind the ability to swap out the transport layer for something custom. It would be nice if I could insert ZeroMQ/NetMQ or some DDS framework in place of WebSockets.

Not sure what that would look like at this point but we won't rule anything out just yet. It might be not pluggable at the SignalR level though.

> Break apart large messages. The current framework allows you to send large messages that clog up the pipe. That should be impossible. It might also be nice to flag a message as high-priority.

That's an interesting one. We'll definitely look at making this possible. I'd prefer if SignalR was decoupled from the idea of message priority. It's supposed to be the last mile, nothing more. The publishing API is too simple to add those types of knobs today.

> The .NET client required me to write my own reflection-based wrapper to make my client-side use the same interface as the server. WampSharp does that automatically, although I still don't like how WampSharp makes you separate methods and events.

We'll look at expanding typed hubs to the client side for outgoing, it doesn't make any sense for receiving messages. We can look at building client side model that allows you to bind a class with methods for the callback case.

> Any exception thrown by the framework needs to be catchable in my code. I still have a variety of SignalR / Websocket exceptions that show up in my unhandled exception catcher.

Which code? There's always exception that you cannot handle.

> It would be nice if the serializer was pluggable the same way the transport layer was. It's handy to use Protobuf-net on occasion.

Yep. We just need to figure out how we add our framing on top of any serialization format and potentially handle different outgoing and incoming formats in an efficient manner.

**Comment by shutdown256** [comment]
I would love to loose the dynamics - or at least make them optional. It looks great in samples but for a large project it is a total mess (IMHO). I kove the idea of being able to define client interface and call client methods trough this interface with full intellisense and strong typing support. Also to be able to access and enumerate users and connections trough LINQ (also as POCO not dynamic). 

Also I would like to be able to define path(s) to hubs javascript, because the default is sometimes incompatible with some client frameworks.

**Comment by MelbourneDeveloper** [comment]
I want to throw in my two cents for binary. It's madness that a framework designed for web sockets which are supposed to be built for performance and terseness do not allow for binary transfer. This shouldn't be a feature, it should have been designed this way from the ground up. String transfer should be a layer based on top of binary not the other way around.

**Comment by muratg** [comment]
Notes from today's design meeting:

# SignalR Core 

## Engineering
  - Make the port build
  - Add tests
  - Port tests

## Testing
  - Implement compatability test (SignalRCore server, SignalR 2.* client)
  - Crossplat testing
  - Perf baseline

## ASP.NET Core idioms
  - Make everything async/await
  - Make restartable client internal
  - DI
  - MVC Application parts?
  - Options/Confiration
  - Auth
  - Cors
  - Logging

## Cut completely
  - Cut hubstate
  - Cut forever frame
  
## Support with the old protocol
  - ... 

## New features
   - Both
		- New protocol version (2)
   - Server
		- Connection state management. Provider based. 
		- JSON-RPC support
		- Binary support
		- Disconnect from server
   - Client
		- Client side hub
		- TypeScript without JQuery, with polyfill
		- C# client rewrite
			- Xamarin client
		- Java client
		- Verify C++ client

**Comment by flightlevel** [comment]
Is this roadmap https://github.com/aspnet/Home/wiki/Roadmap timeframe correct for SignalR?

ie. Q2 2017, if we assume a month or two delay like most projects, we're roughly a year away from release?

**Comment by ryanlin1986** [comment]
Don't know the marketing of java/.Net in USA, but here in China java beat the .Net really badly.
Open source .Net Core is good, but come one please give us a really cake that we can eat.

**Comment by MelbourneDeveloper** [comment]
So by about mid next year, SingalR will be integrated in tp ASP.NET Core? That sounds pretty good.

Wouldn't i be possible to leverage the existing SignalR framework in ASP.NET Core? Or, would port conflicts occur?

**Comment by davidfowl** [comment]
> So by about mid next year, SingalR will be integrated in tp ASP.NET Core? That sounds pretty good.

Yes

> Wouldn't i be possible to leverage the existing SignalR framework in ASP.NET Core? Or, would port conflicts occur?

You can but it doesn't gel with ASP.NET Core. Things like dependency injection, logging etc are not integrated.

**Comment by moozzyk** [comment]
@davidfowl - from what I can see DI and logging is at least partially integrated: 

https://github.com/aspnet/SignalR-Server/search?utf8=%E2%9C%93&q=GetRequiredService&type=Code

https://github.com/aspnet/SignalR-Server/blob/623bd9fdf339da2990e6c75f3dbc6a9e2fa2499f/src/Microsoft.AspNetCore.SignalR.Server/Transports/HttpRequestLifeTime.cs#L23

**Comment by davidfowl** [comment]
@moozzyk the question was about the existing SignalR, not SignalR core or maybe I misread what was meant but "existing SignalR".

**Comment by MelbourneDeveloper** [comment]
> You can but it doesn't gel with ASP.NET Core. Things like dependency injection, logging etc are not integrated.

Just a thing to think about...

It will be great if SignalR is integrated with ASP .NET Core, but at the same time, please be careful not to make us add too many NuGet packages when we just want to create a SignalR project.

**Comment by moozzyk** [comment]
@davidfowl - you are right, sorry - misread the previous comment. 

I would encourage trying the Core version we have now though - we have fixed a few issues in the past couple of weeks and it looks pretty good now (more changes coming though).

**Comment by radu-matei** [comment]
It would also be nice to support custom serialization for camel case. If you change the `JsonSerializer` contract resolver to camel case and register it in the framework it still breaks everything, including the client connection - #115

**Comment by MelbourneDeveloper** [comment]
SignalR should be serialization agnostic if it's not already. It should use dependency injection for serialization.

**Comment by davidfowl** [comment]
If only it were that easy. Don't worry, we're going to be working on that as part of this release.

**Comment by MelbourneDeveloper** [comment]
Why isn't it easy? It should have been designed that way from the ground up.

**Comment by davidfowl** [comment]
Because SignalR has a protocol that wraps user data inside of a bigge payload. We store JSON bytes that represent the user message then wrap the SignalR protocol around it. Today those use the same serializer as the client would need to understand any customizations made to the protocol (camel case, non JSON etc).

**Comment by MelbourneDeveloper** [comment]
Why?

**Comment by davidfowl** [comment]
Not sure what you're asking.

**Comment by MelbourneDeveloper** [comment]
Why is the serialization hard coded?

**Comment by MelbourneDeveloper** [comment]
Obviously, the client and the server will need to use the same type of serialization, but other than that, why does it need to be hard coded?

**Comment by MelbourneDeveloper** [comment]
> Because it was easier to hard code JSON rather than add support to the protocol for client/server serialization pluggability. Something we'll look at this time around as it's involved.

I feel that this is a pretty deep design flaw. I haven't delved in to SignalR deeply enough to really comment yet, but just coming from an outsider's point of view, one of the things I'd be looking for from a web sockets engine is serialization agnosticism. 

If I'm dealing with an external party that wants human readable markup, I will use JSON, but otherwise, I will want to use binary serialization for performance and transfer optimisation.

**Comment by davidfowl** [comment]
Kewl

**Comment by MelbourneDeveloper** [comment]
Serialization should be binary based as that is how data is transferred, and the dependency injection may want to have hooks for encoding and decoding text for markup languages like JSON. But, data should not be transferred as a string at either end. That means that the transfer is forcibly encoded and decoded even if you want to use binary.

**Comment by stajs** [comment]
Like a lot of people, I'll be trying pre-releases out of necessity, and I'm encouraged by what @moozzyk said:

> I would encourage trying the Core version we have now though - we have fixed a few issues in the past couple of weeks and it looks pretty good now (more changes coming though).

However given 1) the churn and risk of regressions at this early stage, and 2) not wanting to add a significant overhead to your support process, I want to raise a suggestion: would you be open to a loose developers :thumbsup:/:thumbsdown: indicator on CI builds? All I want to know is, "yep, this build should be mostly good", or "nope, stay away this one's broken as heck".

**Comment by wesamco** [comment]
Why JS client must not have no dependencies? for the ability to import it with a script tag? 
I think using Reactive Extensions is so natural, (rxjs) for this specific use case SignalR JS client.
I see there is a option of still having a single signalr.js with no deps, also in addition a proper npm dependency for a SignalR js Client with rxjs, thoughts?

**Comment by davidfowl** [comment]
Not going to depend on rxJS in the core. Same reason we don't want to depend on jQuery anymore. It was cool back then, now everybody wants it out. The same might happen with rxJS so we're not baking it into the core.

**Comment by MelbourneDeveloper** [comment]
@davidfowl good thinking!

And the same true of Newtonsoft so best to remove the dependency now before it becomes a problem.

**Comment by wesamco** [comment]
Cool, but I don't agree with classifying rxJS in the same category with jQuery. 
jQuery is a complete JavaScript library used for DOM manipulation etc.. importing it for a project which does not use jQuery was redundant.
but Reactive Extensions is a new development pattern we are seeing in new modern web development frameworks like Angular. it's already here and it's here to stay. we don't go back with development patterns.
but there can still be a SignalR js client npm dependency with rxjs, maybe a 3rd party will create it, maybe I will give it a try in Q1 2017.

**Comment by JanEggers** [comment]
i understand that you dont want to depend on 3rd party in core and i think its a good choice.

but may i ask if u will provide additional bower packages that combine signalr core js with rxJS or angular2 or any.

**Comment by MelbourneDeveloper** [comment]
Wrong. React will just be another bunch of ideas thrown on the garbage heap in a few years. 

REMOVE ALL DEPENDENCIES

**Comment by davidfowl** [comment]
@wesamco There *can* be, but there won't be.

**Comment by davidfowl** [comment]
> but may i ask if u will provide additional bower packages that combine signalr core js with rxJS or angular2 or any.

Sure, there can be a bunch of integration packages eventually.

**Comment by davidfowl** [comment]
@MelbourneDeveloper 

> And the same is true of JSON so best to remove the dependency now before it becomes a problem.

Let it go dude.

**Comment by MelbourneDeveloper** [comment]
Hahaha. 

Well, don't say I didn't mention it down the track when other people complain about it.

**Comment by fhelwanger** [comment]
@MelbourneDeveloper Reactive Extensions != React

**Comment by fhelwanger** [comment]
Anyway, I think it's a good idea to **not** have any dependencies.

If some kind of "glue code" could make integration with RxJS, or any other package, it can be done outside of the core.
